sql/core/src/main/scala/org/apache/spark/sql/Melt.scala

```diff
@@ -0,0 +1,125 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql
+
+import org.apache.spark.sql.catalyst.analysis.{AnsiTypeCoercion, TypeCoercion}
+import org.apache.spark.sql.catalyst.expressions.{AttributeReference, Cast, Expression, Literal}
+import org.apache.spark.sql.catalyst.expressions.objects.AssertNotNull
+import org.apache.spark.sql.catalyst.plans.logical.Expand
+import org.apache.spark.sql.functions.col
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types.{DataType, StringType}
+
+private[sql] object Melt {
+  def of[_](ds: Dataset[_],
+            ids: Seq[String],
+            values: Seq[String] = Seq.empty,
+            dropNulls: Boolean = false,
+            variableColumnName: String = "variable",
+            valueColumnName: String = "value"): DataFrame = {
+    // values should be disjoint to ids
+    if (values.intersect(ids).nonEmpty) {
+      throw new IllegalArgumentException(s"A column cannot be both an id and a value column: " +
```

Comment:
The errors here should leverage `QueryCompilationErrors` I believe. cc @MaxGekk

Reply:
If the error can be triggered by users, please, introduce an error class or re-use an existing one. For example:
https://github.com/apache/spark/blob/625afb4e1aefda59191d79b31f8c94941aedde1e/core/src/main/resources/error/error-classes.json#L75-L78

Reply:
Done in 764d71a4. This adds 5 melt-specific error messages to `error-classes.json`. Notably, it adds a generic derivative of `MISSING_COLUMN`: `MISSING_COLUMNS`. If the code knows about multiple missing columns, this allows you to report all of them, not only one. This enables the user to fix them all and not have to run iteratively into the same exception with different column names. On the downside: if there is only one column, the error talks about columns (plural) while listing only a single column (singular).

@MaxGekk what do you think?